File: backend/app/score.py
"""Deterministic Opportunity Score (ARCH §5). Pure function — no LLM, no I/O."""
from typing import Optional

from .config import SCORE_WEIGHTS as W


def _clamp(v: float, lo: float, hi: float) -> float:
    return min(hi, max(lo, v))


def _r(v: float) -> float:
    return round(v * 10) / 10


def compute_score(
    signals: list, trend_slope: Optional[float], recent_news_count: int, ads_count: int,
    has_priced_competitors: bool, jobs_count: int, gap_status: str, direct_competitors: int,
    weak_competitor_share: float, evidence_count: int, distinct_domains: int, block_type_count: int,
) -> dict:
    domains = {d for s in signals for d in s.get("domains", []) if d}
    avg_intensity = sum(s["intensity"] for s in signals) / len(signals) if signals else 0
    pain = min(1, len(domains) / 6) * (avg_intensity / 5) if signals else 0

    # missing inputs score 0 — never imputed
    momentum = 0.0
    if trend_slope is not None:
        momentum = (_clamp(trend_slope, -0.5, 1.0) + 0.5) / 1.5
    if recent_news_count >= 2:
        momentum = min(1, momentum + 0.2)

    commercial = 0.5 * min(1, ads_count / 4) + 0.3 * (1 if has_priced_competitors else 0) + 0.2 * min(1, jobs_count / 10)

    factor = {"open": 1.0, "partially-served": 0.5}.get(gap_status, 0.1)
    whitespace = factor * max(0, 1 - min(1, direct_competitors / 8))

    parts = {"pain": pain, "momentum": momentum, "commercial": commercial, "whitespace": whitespace, "weakRivals": weak_competitor_share}
    total = sum(W[k] * v for k, v in parts.items())

    if evidence_count >= 30 and distinct_domains >= 4 and block_type_count >= 3:
        confidence = "High"
    elif evidence_count >= 12 and distinct_domains >= 2 and block_type_count >= 2:
        confidence = "Med"
    else:
        confidence = "Low"

    # JS Math.round semantics (half up), so scores match the recorded demo run
    return {
        "total": min(100, int(total + 0.5)),
        "subScores": {k: _r(W[k] * v) for k, v in parts.items()},
        "confidence": confidence,
    }

File: src/lib/pipeline/score.ts
import { CONFIG } from "../config";
import type { Confidence, SubScores } from "../schemas";
import type { GapStatus } from "../schemas";

export interface ScoreInput {
  signals: {
    intensity: number;
    domains: string[];
  }[];
  trendSlope: number | null; // -1..+inf (fraction of baseline)
  recentNewsCount: number;
  adsCount: number;
  hasPricedCompetitors: boolean;
  jobsCount: number;
  gapStatus: GapStatus;
  directCompetitors: number;
  weakCompetitorShare: number; // 0..1
  evidenceCount: number;
  distinctDomains: number;
  blockTypeCount: number;
}

export interface ScoreResult {
  total: number;
  subScores: SubScores;
  confidence: Confidence;
  inputs: ScoreInput;
}

const W = CONFIG.score;

export function computeScore(input: ScoreInput): ScoreResult {
  const pain =
    Math.min(1, input.signals.length === 0 ? 0 : countDomains(input) / 6) *
    (avgIntensity(input) / 5);

  let momentum = 0;
  if (input.trendSlope !== null) {
    const slope = clamp(input.trendSlope, -0.5, 1.0);
    momentum = (slope + 0.5) / 1.5;
  }
  if (input.recentNewsCount >= 2) momentum = Math.min(1, momentum + 0.2);

  const commercial =
    0.5 * Math.min(1, input.adsCount / 4) +
    0.3 * (input.hasPricedCompetitors ? 1 : 0) +
    0.2 * Math.min(1, input.jobsCount / 10);

  const whitespaceFactor =
    input.gapStatus === "open" ? 1.0
    : input.gapStatus === "partially-served" ? 0.5
    : 0.1;
  const crowdPenalty = 1 - Math.min(1, input.directCompetitors / 8);
  const whitespace = whitespaceFactor * Math.max(0, crowdPenalty);

  const weakRivals = input.weakCompetitorShare;

  const weighted =
    W.pain * pain +
    W.momentum * momentum +
    W.commercial * commercial +
    W.whitespace * whitespace +
    W.weakRivals * weakRivals;

  const subScores: SubScores = {
    pain: round(W.pain * pain),
    momentum: round(W.momentum * momentum),
    commercial: round(W.commercial * commercial),
    whitespace: round(W.whitespace * whitespace),
    weakRivals: round(W.weakRivals * weakRivals),
  };

  const confidence = confidenceOf(input);

  return {
    total: Math.min(100, Math.round(weighted)),
    subScores,
    confidence,
    inputs: input,
  };
}

function countDomains(input: ScoreInput): number {
  return new Set(input.signals.flatMap((s) => s.domains).filter(Boolean)).size;
}

function avgIntensity(input: ScoreInput): number {
  const sum = input.signals.reduce((a, s) => a + s.intensity, 0);
  return input.signals.length === 0 ? 0 : sum / input.signals.length;
}

function confidenceOf(input: ScoreInput): Confidence {
  const srcTypes = input.blockTypeCount;
  if (input.evidenceCount >= 30 && input.distinctDomains >= 4 && srcTypes >= 3) {
    return "High";
  }
  if (input.evidenceCount >= 12 && input.distinctDomains >= 2 && srcTypes >= 2) {
    return "Med";
  }
  return "Low";
}

function clamp(v: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, v));
}

function round(v: number): number {
  return Math.round(v * 10) / 10;
}